flink-connector-elasticsearch8/src/test/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBaseITCase.java

```diff
@@ -21,86 +21,144 @@
 
 package org.apache.flink.connector.elasticsearch.sink;
 
+import org.apache.flink.connector.base.sink.writer.ElementConverter;
+import org.apache.flink.testutils.junit.extensions.parameterized.Parameter;
+import org.apache.flink.testutils.junit.extensions.parameterized.ParameterizedTestExtension;
+import org.apache.flink.testutils.junit.extensions.parameterized.Parameters;
+
+import co.elastic.clients.elasticsearch.core.bulk.BulkOperationVariant;
+import co.elastic.clients.elasticsearch.core.bulk.IndexOperation;
 import org.apache.http.HttpHost;
+import org.apache.http.auth.AuthScope;
+import org.apache.http.auth.UsernamePasswordCredentials;
+import org.apache.http.client.CredentialsProvider;
+import org.apache.http.impl.client.BasicCredentialsProvider;
 import org.apache.http.util.EntityUtils;
 import org.elasticsearch.client.Request;
 import org.elasticsearch.client.Response;
 import org.elasticsearch.client.RestClient;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.TestTemplate;
+import org.junit.jupiter.api.extension.ExtendWith;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.testcontainers.containers.output.Slf4jLogConsumer;
+import org.testcontainers.containers.wait.strategy.LogMessageWaitStrategy;
 import org.testcontainers.containers.wait.strategy.Wait;
 import org.testcontainers.elasticsearch.ElasticsearchContainer;
-import org.testcontainers.junit.jupiter.Container;
-import org.testcontainers.junit.jupiter.Testcontainers;
 import org.testcontainers.utility.DockerImageName;
 
 import java.io.IOException;
 import java.time.Duration;
+import java.util.Arrays;
+import java.util.List;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
-/** Base Integration tests class. */
-@Testcontainers
-public class ElasticsearchSinkBaseITCase {
+/**
+ * {@link ElasticsearchSinkBaseITCase} is the base class for integration tests.
+ *
+ * <p>It is extended with the {@link ParameterizedTestExtension} for parameterized testing against
+ * secure and non-secure Elasticsearch clusters. Tests must be annotated by {@link TestTemplate} in
+ * order to be parameterized.
+ *
+ * <p>The cluster is running via test containers. In order to reuse the singleton containers by all
+ * inheriting test classes, we manage their lifecycle. The two containers are started only once when
+ * this class is loaded. At the end of the test suite the Ryuk container that is started by
+ * Testcontainers core will take care of stopping the singleton container.
+ */
+@ExtendWith(ParameterizedTestExtension.class)
```

Comment:
Thanks @liuml07 , I think the usage of 

```suggestion
@ExtendWith(ParameterizedTestExtension.class)
```

is not really bringing value here since the test cases depend on static containers (and  parameterized tests do not support that). However, I think you have quite interesting idea to explore, I would suggest to try that:
 - create 2 containers (as you do now but using the `@Testcontainers` extension):
    ```
    @Container
     public static final ElasticsearchContainer ES_CONTAINER = createElasticsearchContainer();

    @Container
    public static final ElasticsearchContainer ES_SECURE_CONTAINER = createElasticsearchSecureContainer();
    ```
 - consequently, use 2 `RestClient` clients, secure and non-secure
 - introduce abstract method, fe `abstract boolean enableSecurity()` so each test case would have to override and the base test case could use to make the decision which client  + customization to use

WDYT? Thank you.

Reply:
Let me attach a screenshot when I run it locally. Both containers are used by each test case, and also the two containers are used by all test cases of all test classes.

<img width="680" alt="Screenshot 2024-05-29 at 9 59 18 AM" src="https://github.com/apache/flink-connector-elasticsearch/assets/159186/b9886aa9-67b6-43ee-9b8b-941209383268">

Reply:
> The idea of using ParameterizedTestExtension is to make all subclasses run with both secure and non-secure tests by default. So tests do not actually depend on one static container, but instead use both of them.

Oh, apologies, missed that, I think the approach makes perfect sense that.

> I have been reading the doc of testcontainers and didn't find a solution that works best for base classes via @Container managed lifecycle and still efficiently. 

Got it, thank for reference but we still have a problem here since containers are not shutdown properly (we do start them up), that the concerns I have vs using `@Container`, let me try to look into that.

Reply:
Thank you very much for checking again!

This manual lifecycle management for singleton test containers is new to me as well. I presume Ryuk's delay-based container cleanup is not straightforward to many people, so I also put a comment in the base test class.

```
 *
 * <p>The cluster is running via test containers. In order to reuse the singleton containers by all
 * inheriting test classes, we manage their lifecycle. The two containers are started only once when
 * this class is loaded. At the end of the test suite the Ryuk container that is started by
 * Testcontainers core will take care of stopping the singleton container.
 */
```

Reply:
Thank you, @liuml07 !